### PR TITLE
#9080 upgrade IOM model add pins, separate boltassembly boltgrade, ne…

### PR DESCRIPTION
**Pull request by vadura-pavel**
…w boltgrid anchorgrid pingrid

* Did you find critical bug in our code?
* Do you have any new feature you want implemented?
* Did you fix anything?

# Propose PULL Request then and we will examine it

### CODE REVIEW

File: src/IdeaRS.OpenModel/Libraries/Material/Pin.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	/// <summary>
+	/// Bolt assembly
+	/// </summary>
+	[OpenModelClass("CI.StructModel.Libraries.Material.Pin,CI.Material", "CI.StructModel.Libraries.Material.IPin,CI.BasicTypes", typeof(Pin))]
+	[DataContract]
+	[JsonObject(MemberSerialization = MemberSerialization.OptOut)]
+	public class Pin : OpenElementId
+	{
+		/// <summary>
+		/// Name of bolt assembly
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Libraries/Material/Pin.cs:15]: fix summary

File: src/IdeaRS.OpenModel/Libraries/Material/MaterialBoltGrade.cs [hunk 2]
```diff
@@ -10,6 +11,13 @@ namespace IdeaRS.OpenModel.Material
...
 	[DataContract]
 	[JsonObject(MemberSerialization = MemberSerialization.OptOut)]
+	[XmlInclude(typeof(MaterialBoltGradeAISC))]
+	[XmlInclude(typeof(MaterialBoltGradeAUS))]
+	[XmlInclude(typeof(MaterialBoltGradeCISC))]
+	[XmlInclude(typeof(MaterialBoltGradeEc2))]
+	[XmlInclude(typeof(MaterialBoltGradeHKG))]
+	[XmlInclude(typeof(MaterialBoltGradeCHN))]
+	[XmlInclude(typeof(MaterialBoltGradeIND))]
 	public class MaterialBoltGrade : Material
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Libraries/Material/MaterialBoltGrade.cs:21]: should this be abstract now?

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+
+namespace IdeaRS.OpenModel.Connection
+{
+	public class FastenerGridBase : OpenElementId
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs:5]: Need a name parameter

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+		/// <summary>
+		/// Positions of holes in the local coordinate system of the bolt grid
+		/// </summary>
+		public List<IdeaRS.OpenModel.Geometry3D.Point3D> Positions { get; set; }
+
+		/// <summary>
+		/// Id of the connected parts
+		/// </summary>
+		public List<string> ConnectedPartIds { get; set; }
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs:35]: Should this not be references now? not strings.

File: src/IdeaRS.OpenModel/Connection/ConnectionData.cs [hunk 5]
```diff
@@ -15,7 +14,7 @@ public class ConnectionData
 		/// Connection Point Id
 		/// </summary>
 		[DataMember]
-		public int ConenctionPointId { get; set; }
+		public int ConnectionPointId { get; set; }
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/ConnectionData.cs:17]: Finally, Thank god!

File: src/IdeaRS.OpenModel/Libraries/Material/BoltAssembly.cs [hunk 6]
```diff
@@ -127,5 +127,19 @@ public bool WasherAtNut
 		}
 
 		#endregion
+
+		/// <summary>
+		/// Length
+		/// </summary>
+		public double Length
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Libraries/Material/BoltAssembly.cs:134]: This property should be on the BoltGrid?

File: src/IdeaRS.OpenModel/Connection/AnchorGrid.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+﻿using IdeaRS.OpenModel.Parameters;
+
+namespace IdeaRS.OpenModel.Connection
+{
+	/// <summary>
+	/// Data of the anchor grid
+	/// </summary>
+	public class AnchorGrid : FastenerGridBase
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/AnchorGrid.cs:8]: Needs embedment depth property.

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+		/// </summary>
+		public IdeaRS.OpenModel.Geometry3D.Vector3D AxisY { get; set; }
+
+		/// <summary>
+		/// Bolt grid LCS - Axis Z
+		/// </summary>
+		public IdeaRS.OpenModel.Geometry3D.Vector3D AxisZ { get; set; }
+
+		/// <summary>
+		/// Positions of holes in the local coordinate system of the bolt grid
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs:28]: This is wrong description.

**Review by nathanlukeIDEA — CHANGES_REQUESTED** [review]
please review

File: src/IdeaRS.OpenModel/Libraries/Material/MaterialBoltGrade.cs (hunk 2 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Libraries/Material/MaterialBoltGrade.cs:21]: Im not sure this need to be tested during implementation of import.

File: src/IdeaRS.OpenModel/Connection/ConnectionData.cs (hunk 5 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/ConnectionData.cs]: repair

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs (hunk 3 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs:5]: fixed

File: src/IdeaRS.OpenModel/Libraries/Material/Pin.cs (hunk 1 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Libraries/Material/Pin.cs]: fixed

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs (hunk 8 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs]: fixed

File: src/IdeaRS.OpenModel/Connection/AnchorGrid.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/AnchorGrid.cs:8]: done

Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/AnchorGrid.cs:8]: done

File: src/IdeaRS.OpenModel/Libraries/Material/BoltAssembly.cs (hunk 6 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Libraries/Material/BoltAssembly.cs]: removed

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs (hunk 4 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs]: tranformed in to referenced id

File: src/IdeaRS.OpenModel/Libraries/Material/Pin.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		/// <summary>
+		/// Name of pin
+		/// </summary>
+		public string Name { get; set; }
+
+
+		/// <summary>
+		/// Pin Material
+		/// </summary>
+		public ReferenceElement PinMaterial { get; set; }
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Libraries/Material/Pin.cs:23]: Just 'Material'.

File: src/IdeaRS.OpenModel/Connection/PinGrid.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace IdeaRS.OpenModel.Connection
+{
+	/// <summary>
+	/// Data of the pin grid
+	/// </summary>
+	public class PinGrid : FastenerGridBase
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/PinGrid.cs:6]: Need to add 'IsReplaceable' property

File: src/IdeaStatiCa.BimApi/IIdeaGrid.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
+﻿using System.Collections.Generic;
+
+namespace IdeaStatiCa.BimApi
+{
+	/// <summary>
+	/// Represent base of grid
+	/// </summary>
+	public interface IIdeaGrid : IIdeaPersistentObject
```
Comment (nathanlukeIDEA) [inline src/IdeaStatiCa.BimApi/IIdeaGrid.cs:8]: FastenerGrid? to keep consistant

File: src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+		/// <summary>
+		/// List of the connected parts
+		/// </summary>
+		public List<ReferenceElement> ConnectedParts { get; set; }
+
+		/// <summary>
+		/// Assembly
+		/// </summary>
+		public ReferenceElement Assembly { get; set; }
```
Comment (nathanlukeIDEA) [inline src/IdeaRS.OpenModel/Connection/FastenerGridBase.cs:45]: Should we move this to each class to make it more explicit and follow Bimapi implementation?

**Review by nathanlukeIDEA — COMMENTED** [review]
Some small comments

File: src/IdeaStatiCa.BimApi/IIdeaGrid.cs (hunk 11 above)
Comment (vadura-pavel) [reply src/IdeaStatiCa.BimApi/IIdeaGrid.cs:8]: renamed

File: src/IdeaRS.OpenModel/Libraries/Material/Pin.cs (hunk 9 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Libraries/Material/Pin.cs:23]: renamed

File: src/IdeaRS.OpenModel/Connection/PinGrid.cs (hunk 10 above)
Comment (vadura-pavel) [reply src/IdeaRS.OpenModel/Connection/PinGrid.cs:6]: added

**Review by nathanlukeIDEA — APPROVED** [review]

**Review by nathanlukeIDEA — APPROVED** [review]

**Pull request merged**